### ISSUE TITLE
Object of class UnwindExit could not be converted to string

### ISSUE THREAD

**Issue by nunoperalta** [googleads/googleads-php-lib#733]
I wonder if this issue is due to PHP 8.1, but I'm not seeing anywhere mentioned that PHP 8 isn't supported.

I'm using:

		"google/apiclient": "v2.12.1",
		"google/apiclient-services": "v0.229.0",
		"google/auth": "v1.18.0",
		"googleads/googleads-php-lib": "55.0.0",

```
use Google\AdsApi\AdManager\AdManagerServices;
use Google\AdsApi\AdManager\AdManagerSessionBuilder;
use Google\AdsApi\AdManager\Util\v202111\AdManagerDateTimes;
use Google\AdsApi\AdManager\Util\v202111\ReportDownloader;
use Google\AdsApi\AdManager\Util\v202111\StatementBuilder;
use Google\AdsApi\AdManager\v202111\DateRangeType;
use Google\AdsApi\AdManager\v202111\ExportFormat;
use Google\AdsApi\AdManager\v202111\ReportJob;
use Google\AdsApi\AdManager\v202111\ReportService;
use Google\AdsApi\Common\OAuth2TokenBuilder;
```

And getting this crash, fetching Reports from AdManager:

> PHP Fatal error [8192]: **Return type of Google\AdsApi\Common\AdsSoapClient::__soapCall($function_name, $arguments, $options = null, $input_headers = null, &$output_headers = null) should either be compatible with SoapClient::__soapCall(string $name, array $args, ?array $options = null, $inputHeaders = null, &$outputHeaders = null): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice**
> Stack trace:
> #0 ..../vendor/composer/ClassLoader.php(571)
> #1 ..../vendor/composer/ClassLoader.php(571): include()
> #2 ..../vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile('..../vendor/composer/../googleads/googleads-php-lib/src/Google/AdsApi/Common/AdsSoapClient.php')
> #3 ..../vendor/googleads/googleads-php-lib/src/Google/AdsApi/AdManager/v202111/ReportService.php(9): Composer\Autoload\ClassLoader->loadClass('Google\\AdsApi\\Common\\AdsSoapClient')
> #4 ..../vendor/composer/ClassLoader.php(571): include('..../vendor/googleads/googleads-php-lib/src/Google/AdsApi/AdManager/v202111/ReportService.php')
> #5 ..../vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile('..../vendor/composer/../googleads/googleads-php-lib/src/Google/AdsApi/AdManager/v202111/ReportService.php')
> #6 [internal function]: Composer\Autoload\ClassLoader->loadClass('Google\\AdsApi\\AdManager\\v202111\\ReportService')
> #7 ..../vendor/googleads/googleads-php-lib/src/Google/AdsApi/AdManager/AdManagerServiceDescriptor.php(44): ReflectionClass->__construct('Google\\AdsApi\\AdManager\\v202111\\ReportService')
> #8 ..../vendor/googleads/googleads-php-lib/src/Google/AdsApi/AdManager/AdManagerServices.php(60): Google\AdsApi\AdManager\AdManagerServiceDescriptor->__construct('Google\\AdsApi\\AdManager\\v202111\\ReportService')
> #9 myscript.php(48): Google\AdsApi\AdManager\AdManagerServices->get(Object(Google\AdsApi\AdManager\AdManagerSession), 'Google\\AdsApi\\AdManager\\v202111\\ReportService')

...

> PHP Fatal error:  **Object of class UnwindExit could not be converted to string** in ..../vendor/googleads/googleads-php-lib/src/Google/AdsApi/Common/AdsSoapClient.php on line 32
>

**Comment by nunoperalta** [comment]
I was able to fix this by adding return types to these functions:

```
    /**
     * @see SoapClient::__doRequest
     */
    public function __doRequest(
        $request,
        $location,
        $action,
        $version,
        $one_way = 0
    ) : ?string {
        $request = SoapRequests::replaceReferences($request);
        $response = parent::__doRequest($request, $location, $action, $version, $one_way);
        $this->__last_request = $request;

        return $response;
    }

    /**
     * @see SoapClient::__soapCall
     */
    public function __soapCall(
        $function_name,
        $arguments,
        $options = null,
        $input_headers = null,
        &$output_headers = null
    ) : mixed {
```

**Comment by thangduo** [comment]
Thanks nunoperalta for reporting the issue and suggesting a fix. After reading the C source of the `SoapClient::__doRequest` function, it makes sense to add `?string|void` type hint for the return value.

Can you please confirm whether adding the type hint above to just the `SoapClient::__doRequest` function would be sufficient for fixing your issue? I want to avoid adding type hints for the `SoapClient::__soapCall` and other functions without fully understanding their implementation and scope.

Thanks,
Thang

**Comment by nunoperalta** [comment]
Thank you. I had to add to both, as fixing one, another error will occur, so both had to be fixed.
Alternatively, for compatibility with previous PHP versions, you might just want to add the attribute suggested?

> or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice

**Comment by thangduo** [comment]
Thanks for the pointer to the [new feature](https://php.watch/versions/8.1/ReturnTypeWillChange) `#[\ReturnTypeWillChange]`. Unfortunately, that would require upgrading the minimum PHP version to 8.1.
We will keep the current minimum PHP version at 7.4 in 10 months until it reaches its end of life.

I will add the type hints in the next release of this library.

Thanks,
Thang

**Comment by nunoperalta** [comment]
I think that if you go with the suggestion of adding the return types, that's what requires you to update to 8.1  (because union types are not supported in earlier versions)

But if you simply put the attribute, I assume that would be compatible, no?
Earlier versions will simply see it as a comment.

**Comment by PierrickVoulet** [comment]
@thangduo Did you have the opportunity to release a fix for this just yet?

**Comment by PierrickVoulet** [comment]
Closing this because it was fixed with version [v58.0.0](https://github.com/googleads/googleads-php-lib/releases/tag/58.0.0).
